### PR TITLE
Add `init_weights` method to `FlaxMixin`

### PR DESCRIPTION
**Pull request by mishig25**
### Implementation of `init_weights` method is required for any class that is inheriting from [FlaxModelMixin](https://github.com/huggingface/diffusers/blob/32c2be514bb4d3bd76b23a5ceed7865bc34f821d/src/diffusers/modeling_flax_utils.py#L39).

[Here](https://github.com/mishig25/stable-diffusion-jax/blob/70f384f1dcd6f7ea11a10cd5e48e69722fb38990/stable_diffusion_jax/modeling_unet2d.py#L689-L699) is an example:
```py
class UNet2D(nn.Module, FlaxModelMixin, ConfigMixin):
    def init_weights(self, rng: jax.random.PRNGKey) -> FrozenDict:
        # init input tensors
        sample_shape = (1, self.config.sample_size, self.config.sample_size, self.config.in_channels)
        sample = jnp.zeros(sample_shape, dtype=jnp.float32)
        timestpes = jnp.ones((1,), dtype=jnp.int32)
        encoder_hidden_states = jnp.zeros((1, 1, self.config.cross_attention_dim), dtype=jnp.float32)

        params_rng, dropout_rng = jax.random.split(rng)
        rngs = {"params": params_rng, "dropout": dropout_rng}

        return self.init(rngs, sample, timestpes, encoder_hidden_states)["params"]
```

Unlike [transformers.FlaxPretrainedModel.init_weights](https://github.com/huggingface/transformers/blob/bce36ee065f7749c997be4c30a3f9279df7d5dba/src/transformers/modeling_flax_utils.py#L229), [diffusers.FlaxModelMixin.init_weights](https://github.com/huggingface/transformers/blob/bce36ee065f7749c997be4c30a3f9279df7d5dba/src/transformers/modeling_flax_utils.py#L229) signature does not have parameter `input_shape`. Read more [here](https://github.com/huggingface/diffusers/pull/513#discussion_r971271647) & [here](https://github.com/huggingface/diffusers/pull/493#issuecomment-1246550570) on why the decision was made

Users will have 2 option to init weights:
```py
class FlaxModel(nn.Module, FlaxModelMixin, ConfigMixin):
   ...

my_model = FlaxModel()
# option1 (FlaxModelMixin)
params = my_model.init_weights(key)
# option2 (linen.Module)
x = random.normal(key1, (...)) # Dummy input
params = FlaxModel.init(key, x)

# option1 is more convenient since the random input is automatically handled inside `init_weights`, but they will still have an option2 if they want to use it as a normal linen.Module
```

### CODE REVIEW

File: src/diffusers/modeling_flax_utils.py [hunk 1]
```diff
@@ -280,6 +284,7 @@ def from_pretrained(
         revision = kwargs.pop("revision", None)
         from_auto_class = kwargs.pop("_from_auto", False)
         subfolder = kwargs.pop("subfolder", None)
+        prng_key = kwargs.pop("prng_key", None)
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_flax_utils.py:287]: @patil-suraj - this isn't really needed no? We're only inferring the shape here - so maybe let's also use `jax.random.PRNGKey(0)` no?

File: src/diffusers/modeling_flax_utils.py [hunk 2]
```diff
@@ -394,6 +399,82 @@ def from_pretrained(
         # flatten dicts
         state = flatten_dict(state)
 
+        prng_key = prng_key if prng_key is not None else jax.random.PRNGKey(0)
+        params_shape_tree = jax.eval_shape(model.init_weights, prng_key)
+        required_params = set(flatten_dict(unfreeze(params_shape_tree)).keys())
+
+        random_state = flatten_dict(unfreeze(params_shape_tree))
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_flax_utils.py:406]:
```suggestion
        shape_state = flatten_dict(unfreeze(params_shape_tree))
```

File: src/diffusers/modeling_flax_utils.py [hunk 3]
```diff
@@ -272,6 +275,7 @@ def from_pretrained(
         ```"""
         config = kwargs.pop("config", None)
         cache_dir = kwargs.pop("cache_dir", DIFFUSERS_CACHE)
+        ignore_mismatched_sizes = kwargs.pop("ignore_mismatched_sizes", False)
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_flax_utils.py:278]: Can we remove this for now? I don't think we should allow mismatched sizes for now

File: src/diffusers/modeling_flax_utils.py [hunk 4]
```diff
@@ -183,6 +183,9 @@ def to_fp16(self, params: Union[Dict, FrozenDict], mask: Any = None):
         ```"""
         return self._cast_floating_to(params, jnp.float16, mask)
 
+    def init_weights(self, rng: jax.random.PRNGKey) -> Dict:
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_flax_utils.py:186]: I like that we don't allow the `input_shape` to be passed for now since it's much more restricted than Transformers, *i.e.* we should for now always be able to infer the correct shape from the config. This looks good to me!

**Review by patrickvonplaten — COMMENTED** [review]
Just some minor comments regarding naming, and think we should remove the "allow mismatched keys" functionality for now. But apart from this this is top!

File: src/diffusers/modeling_flax_utils.py [hunk 5]
```diff
@@ -394,6 +399,82 @@ def from_pretrained(
...
+                if ignore_mismatched_sizes:
+                    mismatched_keys.append((key, state[key].shape, random_state[key].shape))
+                    state[key] = random_state[key]
+                else:
+                    raise ValueError(
+                        f"Trying to load the pretrained weight for {key} failed: checkpoint has shape "
+                        f"{state[key].shape} which is incompatible with the model shape {random_state[key].shape}. "
+                        "Using `ignore_mismatched_sizes=True` if you really want to load this checkpoint inside this "
+                        "model."
+                    )
```
Comment (pcuenca) [inline src/diffusers/modeling_flax_utils.py:432]: Just renaming as suggested above.

```suggestion
            if key in shape_state and state[key].shape != shape_state[key].shape:
                if ignore_mismatched_sizes:
                    mismatched_keys.append((key, state[key].shape, shape_state[key].shape))
                    state[key] = shape_state[key]
                else:
                    raise ValueError(
                        f"Trying to load the pretrained weight for {key} failed: checkpoint has shape "
                        f"{state[key].shape} which is incompatible with the model shape {shape_state[key].shape}. "
                        "Using `ignore_mismatched_sizes=True` if you really want to load this checkpoint inside this "
                        "model."
                    )
```

**Comment by pcuenca** [comment]
I agree with @patrickvonplaten's comments. This is very cool, I'll test it later!

File: src/diffusers/modeling_flax_utils.py (hunk 5 above)
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py]: handled in https://github.com/huggingface/diffusers/pull/513/commits/305a5448d1de9c2b4c45ee47b150d37f6919466f

File: src/diffusers/modeling_flax_utils.py (hunk 2 above)
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py]: handled in https://github.com/huggingface/diffusers/pull/513/commits/305a5448d1de9c2b4c45ee47b150d37f6919466f

File: src/diffusers/modeling_flax_utils.py (hunk 1 above)
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py]: you are right! handled in https://github.com/huggingface/diffusers/pull/513/commits/803da8f666b2f17c6accf4a73c804fdefd459461

File: src/diffusers/modeling_flax_utils.py (hunk 3 above)
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py]: made it stricter https://github.com/huggingface/diffusers/pull/513/commits/73e0bc692c5761e55faff39c80a26d7a3cfc748c

**Comment by mishig25** [comment]
should be ready for review, also updated the description with more details for clairfication

**Review by pcuenca — APPROVED** [review]
Looks great!

File: src/diffusers/modeling_flax_utils.py (hunk 4 above)
Comment (patil-suraj) [reply src/diffusers/modeling_flax_utils.py:186]: +1

File: src/diffusers/modeling_flax_utils.py [hunk 6]
```diff
@@ -183,6 +183,9 @@ def to_fp16(self, params: Union[Dict, FrozenDict], mask: Any = None):
         ```"""
         return self._cast_floating_to(params, jnp.float16, mask)
 
+    def init_weights(self, rng: jax.random.PRNGKey) -> Dict:
+        raise NotImplementedError(f"init method has to be implemented for {self}")
```
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:187]:
```suggestion
        raise NotImplementedError(f"init_weights method has to be implemented for {self}")
```

File: src/diffusers/modeling_flax_utils.py [hunk 7]
```diff
@@ -394,6 +393,76 @@ def from_pretrained(
         # flatten dicts
         state = flatten_dict(state)
 
+        prng_key = jax.random.PRNGKey(0)
+        params_shape_tree = jax.eval_shape(model.init_weights, prng_key)
```
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:397]: If this is not used anywhere down in the code
```suggestion
        params_shape_tree = jax.eval_shape(model.init_weights, rng=jax.random.PRNGKey(0))
```

File: src/diffusers/modeling_flax_utils.py [hunk 8]
```diff
@@ -394,6 +393,76 @@ def from_pretrained(
...
+
+        if len(unexpected_keys) > 0:
+            logger.warning(
+                f"Some weights of the model checkpoint at {pretrained_model_name_or_path} were not used when"
+                f" initializing {model.__class__.__name__}: {unexpected_keys}\n- This IS expected if you are"
+                f" initializing {model.__class__.__name__} from the checkpoint of a model trained on another task or"
+                " with another architecture (e.g. initializing a BertForSequenceClassification model from a"
+                " BertForPreTraining model).\n- This IS NOT expected if you are initializing"
+                f" {model.__class__.__name__} from the checkpoint of a model that you expect to be exactly identical"
+                " (initializing a BertForSequenceClassification model from a BertForSequenceClassification model)."
```
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:434]: docstring should be updated for `diffusers`

**Review by patil-suraj — APPROVED** [review]
Very cool! Just left couple of nits

File: src/diffusers/modeling_flax_utils.py (hunk 8 above)
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py]: wdyt https://github.com/huggingface/diffusers/pull/513/commits/869014ca901aabd82b7e053feb251caed74f9484 since we don't have LLM-like heads in diffusion models

**Comment by patrickvonplaten** [comment]
Merging as tests are taking too long at the moment
